kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/resolver/ClusterResolutionResult.java

```diff
@@ -6,109 +6,197 @@
+    }
+
+    /**
+     * Find all dangling references (recursive) from this VirtualKafkaCluster
+     * @return a stream of all dangling references
+     */
+    public Stream<ClusterResolutionResult.DanglingReference> findDanglingReferences() {
+        return allReferentResolutionResults().filter(ResolutionResult::dangling)
+                .map(resolutionResult1 -> new ClusterResolutionResult.DanglingReference(resolutionResult1.referrer(), resolutionResult1.reference()));
+    }
+
+    /**
+     * Find all dangling references (recursive) from the specified referrer to the specified kind
+     * @param fromReferrer a LocalRef identifying the Referrer
+     * @param toKind a Referent kind
+     * @return a stream of all dangling references from the specified LocalRef to the specified kind
+     */
     @SuppressWarnings("java:S1452")
-    public Stream<LocalRef<?>> findDanglingReferences(LocalRef<?> from, String kindTo) {
-        Objects.requireNonNull(from);
-        Objects.requireNonNull(kindTo);
-        return danglingReferences.stream().filter(r -> r.from().equals(from) && kindTo.equals(r.to.getKind())).map(DanglingReference::to);
+    public Stream<LocalRef<?>> findDanglingReferences(LocalRef<?> fromReferrer, String toKind) {
+        Objects.requireNonNull(fromReferrer);
+        Objects.requireNonNull(toKind);
+        return findDanglingReferences().filter(r -> r.from().equals(fromReferrer) && toKind.equals(r.to().getKind())).map(
+                ClusterResolutionResult.DanglingReference::to);
     }
 
+    /**
+     * Find all dangling references (recursive) from the specified fromKind to the specified toKind
+     * @param fromKind a Referrer kind
+     * @param toKind a Referent kind
+     * @return a stream of all dangling references from the specified fromKind to the specified toKind
+     */
     // using wildcards intentionally
     @SuppressWarnings("java:S1452")
     public Stream<LocalRef<?>> findDanglingReferences(String fromKind, String toKind) {
         Objects.requireNonNull(fromKind);
         Objects.requireNonNull(toKind);
-        return danglingReferences.stream().filter(r -> fromKind.equals(r.from().getKind()) && toKind.equals(r.to().getKind()))
-                .map(DanglingReference::to);
+        return findDanglingReferences().filter(r -> fromKind.equals(r.from().getKind()) && toKind.equals(r.to().getKind()))
+                .map(ClusterResolutionResult.DanglingReference::to);
     }
 
     /**
-     * A result is fully resolved if:
-     * <ol>
-     *     <li>All references can be retrieved from kubernetes (we have no dangling refs)</li>
-     *     <li>No referent has a ResolvedRefs=False condition, which declares that it has unresolved dependencies</li>
-     * </ol>
+     * Find referents and cluster with stale status (status.observedGeneration != metadata.generation)
+     * @return LocalRefs to referents and cluster with stale status
```

Comment:
I don't understand what "referents and cluster" means, given this returns some `LocalRefs` (==referents). Where does the cluster part come from?

Reply:
`referents` means referents of the cluster. the cluster itself isn't a Referent but we want to treat it the same way sometimes. The ProxyReconciler want to ask "do any of the referents or the cluster have ResolvedRefs=false?". Where the VKC Reconciler only wants to ask "do any of the referents have ResolvedRefs=false?"

Reply:
> we want to treat it the same way sometimes

That is not at all obvious, and is unintuitive for anyone with the CR model in their head (where clusters do not refer to clusters).
* Is it really necessary? I mean, can't the `ProxyReconciler` just make two calls? Or could it concatenate a stream of referents with a LocalRef for the cluster at the call site which want to tread things that way?
* If it is really necessary then you should at least explain this in the Javadoc.

Reply:
Okay, I'll shuffle it around so we expose a couple of boolean methods and have the clients handle the concatenation